frontend/packages/dev-console/src/components/pipelineruns/logs/logs-utils.ts

```diff
@@ -0,0 +1,142 @@
+import { saveAs } from 'file-saver';
+import {
+  PodKind,
+  ContainerSpec,
+  ContainerStatus,
+  resourceURL,
+  k8sGet,
+} from '@console/internal/module/k8s';
+import { LOG_SOURCE_TERMINATED, LineBuffer } from '@console/internal/components/utils';
+import { PodModel } from '@console/internal/models';
+import { coFetchText } from '@console/internal/co-fetch';
+import { errorModal } from '@console/internal/components/modals';
+import { containerToLogSourceStatus } from '../../../utils/pipeline-utils';
+import { TaskRuns } from '../../../utils/pipeline-augment';
+
+const getSortedContainerStatus = (
+  containers: ContainerSpec[],
+  containerStatuses: ContainerStatus[],
+): ContainerStatus[] => {
+  const containerNames = containers.map((c) => c.name);
+  const sortedContainerStatus = [];
+  containerStatuses.forEach((cs) => {
+    const containerIndex = containerNames.indexOf(cs.name);
+    sortedContainerStatus[containerIndex] = cs;
+  });
+  return sortedContainerStatus;
+};
+
+export const getRenderContainers = (
+  pod: PodKind,
+): { containers: ContainerSpec[]; stillFetching: boolean } => {
+  const containers: ContainerSpec[] = pod.spec?.containers ?? [];
+  const containerStatuses: ContainerStatus[] = pod.status?.containerStatuses ?? [];
+
+  const sortedContainerStatuses = getSortedContainerStatus(containers, containerStatuses);
+
+  const firstRunningCont = sortedContainerStatuses.findIndex(
+    (container) => containerToLogSourceStatus(container) !== LOG_SOURCE_TERMINATED,
+  );
+  return {
+    containers: containers.slice(
+      0,
+      firstRunningCont === -1 ? containers.length : firstRunningCont + 1,
+    ),
+    stillFetching: firstRunningCont !== -1,
+  };
+};
+
+const getOrderedStepsFromPod = (name: string, ns: string): Promise<ContainerStatus[]> => {
+  return k8sGet(PodModel, name, ns)
+    .then((pod: PodKind) => {
+      return getSortedContainerStatus(
+        pod.spec.containers ?? [],
+        pod.status?.containerStatuses ?? [],
+      );
+    })
+    .catch((err) => {
+      errorModal({ error: err.message || 'Error downloading logs.' });
+      return [];
+    });
+};
+
+type StepsWatchUrl = {
+  [key: string]: {
+    name: string;
+    steps: { [step: string]: string };
```

Comment:
```suggestion
    steps: { [step: string]: WatchURLStatus };
```
```
type WatchURLStatus = {
  status: string;
  url: string;
};
```
Needed for my next comment.